# chunker: default program

In [3]:
from default_edit import *
import os

## Run the default solution on dev

In [4]:
chunker = LSTMTagger(os.path.join('data', 'train.txt.gz'), os.path.join('data', 'chunker'), '.tar')
decoder_output = chunker.decode('data/input/dev.txt')

100%|██████████| 1027/1027 [00:07<00:00, 145.65it/s]


## Evaluate the default output

In [5]:
flat_output = [ output for sent in decoder_output for output in sent ]
import conlleval
true_seqs = []
with open(os.path.join('data','reference','dev.out')) as r:
    for sent in conlleval.read_file(r):
        true_seqs += sent.split()
conlleval.evaluate(true_seqs, flat_output)

processed 23663 tokens with 11896 phrases; found: 12124 phrases; correct: 9535.
accuracy:  88.16%; (non-O)
accuracy:  89.03%; precision:  78.65%; recall:  80.15%; FB1:  79.39
             ADJP: precision:  47.47%; recall:  33.19%; FB1:  39.06  158
             ADVP: precision:  65.49%; recall:  55.78%; FB1:  60.24  339
            CONJP: precision:   0.00%; recall:   0.00%; FB1:   0.00  2
             INTJ: precision:   0.00%; recall:   0.00%; FB1:   0.00  0
               NP: precision:  77.97%; recall:  82.81%; FB1:  80.32  6624
               PP: precision:  93.41%; recall:  87.71%; FB1:  90.47  2292
              PRT: precision:  71.79%; recall:  62.22%; FB1:  66.67  39
             SBAR: precision:  69.19%; recall:  54.01%; FB1:  60.66  185
               VP: precision:  71.47%; recall:  77.08%; FB1:  74.17  2485


(78.64566149785549, 80.15299260255549, 79.39217318900916)

## Documentation

Write some beautiful documentation of your program here.

## Analysis

Do some analysis of the results. What ideas did you try? What worked and what did not?

In [7]:
from default_edit import *
import os
chunker = LSTMTagger(os.path.join('data', 'train.txt.gz'), os.path.join('data', 'chunker'), '.tar')
print(chunker.model)


LSTMTaggerModel(
  (word_embeddings): Embedding(9675, 128)
  (lstm): LSTM(428, 64)
  (hidden2tag): Linear(in_features=64, out_features=22, bias=True)
)


In [16]:
trainfile = os.path.join('data', 'train.txt.gz')
modelfile = os.path.join('data', 'chunker')
modelsuffix = '.tar'
unk = '[UNK]'

In [17]:
chunker = LSTMTagger(trainfile, modelfile, modelsuffix, unk)

In [18]:
chunker.word_to_ix
#chunker.tag_to_ix  
#chunker.ix_to_tag 

{'Tiger': 9500,
 'age': 4105,
 'Atlas': 6985,
 'allegiance': 9550,
 'violation': 1425,
 'fiberglass': 8691,
 '1.0': 3414,
 'Safeco': 7898,
 'things': 3282,
 'enough': 1578,
 'lesser': 7819,
 'tax-exempt': 1626,
 'Lone': 8443,
 'Plaza': 6984,
 'Ballhaus': 9415,
 'admission': 3247,
 'commercials': 6457,
 'jacket': 6405,
 'overruns': 7630,
 'decide': 1053,
 'stand': 1552,
 'October': 149,
 'restaurant': 2220,
 'Guinness': 9213,
 'wars': 3719,
 'widow': 5556,
 'LONDON': 8735,
 '2005': 7922,
 'Stock-index': 5283,
 'dilutive': 8958,
 'bleak': 5264,
 'other': 309,
 'VCR': 5530,
 'positions': 224,
 'management-led': 1238,
 'merchant': 9326,
 'jetliner': 3092,
 'reforming': 7408,
 'hard': 3299,
 'through': 849,
 'denies': 4414,
 'Golomb': 7984,
 'Real-estate': 7415,
 'Roe': 7883,
 'telecommunications': 3152,
 'Microsoft': 7307,
 'Industrial': 360,
 'separate': 5221,
 'third-party': 2635,
 'Ministry': 3121,
 'no-load': 5223,
 'bureau': 4133,
 'Clearing': 8982,
 'struggling': 3735,
 'Stevens': 58

In [6]:
training_data = []

In [9]:
with gzip.open(trainfile, 'rt') as f:
                training_data = read_conll(f)

In [17]:
training_data

[(('[UNK]',
   'in',
   'the',
   'pound',
   'is',
   'widely',
   'expected',
   'to',
   'take',
   'another',
   'sharp',
   'dive',
   'if',
   'trade',
   'figures',
   'for',
   'September',
   ',',
   'due',
   'for',
   'release',
   'tomorrow',
   ',',
   'fail',
   'to',
   'show',
   'a',
   'substantial',
   'improvement',
   'from',
   'July',
   'and',
   'August',
   "'s",
   'near-record',
   '[UNK]',
   '.'),
  ('B-NP',
   'B-PP',
   'B-NP',
   'I-NP',
   'B-VP',
   'I-VP',
   'I-VP',
   'I-VP',
   'I-VP',
   'B-NP',
   'I-NP',
   'I-NP',
   'B-SBAR',
   'B-NP',
   'I-NP',
   'B-PP',
   'B-NP',
   'O',
   'B-ADJP',
   'B-PP',
   'B-NP',
   'B-NP',
   'O',
   'B-VP',
   'I-VP',
   'I-VP',
   'B-NP',
   'I-NP',
   'I-NP',
   'B-PP',
   'B-NP',
   'I-NP',
   'I-NP',
   'B-NP',
   'I-NP',
   'I-NP',
   'O')),
 (('Chancellor',
   'of',
   'the',
   'Exchequer',
   'Nigel',
   'Lawson',
   "'s",
   'restated',
   'commitment',
   'to',
   'a',
   'firm',
   'monetary',
   '

In [50]:
# The 3 vectors that we have to create, refer to the research paper, I am creating dict 
# initially but will convert to vector and tensor eventually to feed into the model
beginning = {}
internal = {}
ending = {}

for char in string.printable:
    beginning[char] = 0
    internal [char] = 0
    ending[char] = 0
    

In [51]:
# creating 3 vectors for the word "hello"
# notice that we are getting different indices from what is given on the homework page,
# I don't think it will matter at the end
word = "hello"
length = len(word)
b = word[0]
e = word [length-1]
for char in word[1:length-1]:
    internal[char] = internal.get(char) + 1
beginning[b] = beginning.get(b) + 1
ending[e] = ending.get(e) + 1


In [95]:
beginning_vect = []

k = list(beginning.keys())
k.sort()
for i in k:
    beginning_vect.append(beginning[i])
    
internal_vect = []    

k = list(internal.keys())
k.sort()
for i in k:
    internal_vect.append(internal[i])

    
ending_vect = []

k = list(ending.keys())
k.sort()
for i in k:
    ending_vect.append(ending[i])


In [103]:
# Converting the vectors into tensors.
beginning_tensor = torch.FloatTensor(beginning_vect)
internal_tensor = torch.FloatTensor(internal_vect)
ending_tensor = torch.FloatTensor(ending_vect)

In [104]:
# when the instruction says we have to make 2-d tensor of size, sentence length times 300
# It just means concatenating them together.
# this can become 2-d when we go over various words, note here we are only doing for "hello"

character_level = torch.cat((beginning_tensor, internal_tensor, ending_tensor), 0)

In [129]:
character_level

tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.,
        0., 2., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 

In [130]:
character_level.shape

torch.Size([300])

In [115]:
# An example of how we can create a 2-d tensor, we can just stack tensors of length 300
# for various words in a sentence
# Don't take this example literally since we are just stacking beginning_tensor
# internal_tensor
stacktensor = torch.stack([beginning_tensor, internal_tensor], dim = 0)

In [117]:
stacktensor

tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 1., 0., 0., 0., 0., 0., 0., 2., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])

In [118]:
stacktensor.shape

torch.Size([2, 100])

In [ ]:
# Can someone explain how are the word embeddings (from default.py) which is a 
#2D tensor of size, sentence length times 128
# If we can figure this out, we just have to train

In [3]:
model = LSTMTaggerModel(128, 64, 9675, 22)

In [4]:
model.word_embeddings

Embedding(9675, 128)

In [11]:
for sent, tags in training_data:
    dummy_sent = sent

In [12]:
sentence = dummy_sent

In [13]:
sentence

('It',
 'is',
 'also',
 'pulling',
 '20',
 'people',
 'out',
 'of',
 'Puerto',
 'Rico',
 ',',
 'who',
 'were',
 'helping',
 '[UNK]',
 'Hugo',
 'victims',
 ',',
 'and',
 'sending',
 'them',
 'to',
 'San',
 'Francisco',
 'instead',
 '.')

In [14]:
model.forward(sentence)

TypeError: embedding(): argument 'indices' (position 2) must be Tensor, not tuple

In [15]:
model = LSTMTagger

In [22]:
for sentence, tags in tqdm.tqdm(training_data):
                # Step 2. Get our inputs ready for the network, that is, turn them into
                # Tensors of word indices.
                sentence_in = prepare_sequence(sentence, chunker.word_to_ix, unk)
                
                targets = prepare_sequence(tags,chunker.tag_to_ix,unk)



100%|██████████| 8936/8936 [00:00<00:00, 31082.32it/s]


In [25]:
sentence_in.shape

torch.Size([26])

In [26]:
targets.shape

torch.Size([26])

In [27]:
targets

tensor([0, 3, 4, 4, 0, 2, 1, 1, 0, 2, 6, 0, 3, 4, 0, 2, 2, 6, 6, 3, 0, 1, 0, 2,
        8, 6])

# Analysis

The syntax of a natural language involves the arrangement of tokens into meaningful groups. Phrasal chunking is the task of finding non-recursive syntactic groups of words. For example, the sentence:
<br>


"The current debt will rise to 1.8 billion in December"

The pharasal chunks for this will be:
<br>
<br>
<b>
[NP the current debt] [VP will rise] [PP to] [NP 1.8 billion] [PP in] [NP December] 
</b>
<br>
<br>
Besides doing phrasal chunking we also have to handle data which have been infected with noise. 

We build a semi-character Recurrent Neural Network as described by the paper titled "Robsut Wrod Reocginiton via Semi-Character Recurrent Neural Network".

The idea presented in this paper is inspired from the <I>Cmabrigde Uinervtisy</I> effect where the word recognition model is based on semi-character level recurrent neural network

The schematic illustraion of semi-character recurrent neural network is as follows:
<br>
<br>
<img src="img.png">
<br>
<br>
The following method was used to create a character level representation of the word:
1. A one-hot vector "beginning" for the first character of the word.
2. A vector "internal" where the index of a character has the count of that character in the word.
3. A one-hot vector "ending" for the last character of the word.

Eventually we concatenated to the word embedding input to the chunker RNN an input vector that is the character level representation of the word.

This gave us a dev.out score of 77.4683

We tried to improve this method by utilizing a better optimizer. Our original solution used Stochastic Gradient Descent. We had an intuition that it may not be converging quickly enough and led to suboptimal results. 

Figure comparing gradient descent v/s gradient descent with momentum.
<br>
<br>
<img src="img1.png">
<br>
<br>

Looking at the figure we can see that using momentum can lead to faster convergence.

We tried the following optimizers:

1. Adam
2. Stochastic Gradient Descent with momentum

We got the following results:

| Optimizer | Dev.out Score |
| --- | --- |
| SGD | 77.4683 |
| Adam | ~75 |
| SGD with momentum | 79.3922 |
| RMSprop | 78.6700 |

We can see that we have got a significant boost in the dev.out score using SGD with momentum and RMSprop. 